lib/AST/ArchetypeBuilder.cpp

```diff
@@ -467,13 +468,46 @@ auto ArchetypeBuilder::PotentialArchetype::getNestedType(
   // archetype conforms.
   for (auto &conforms : ConformsTo) {
     for (auto member : conforms.first->lookupDirect(nestedName)) {
-      auto assocType = dyn_cast<AssociatedTypeDecl>(member);
-      if (!assocType)
+      PotentialArchetype *pa;
+      
+      if (auto assocType = dyn_cast<AssociatedTypeDecl>(member)) {
+        // Resolve this nested type to this associated type.
+        pa = new PotentialArchetype(this, assocType);
+      } else if (auto alias = dyn_cast<TypeAliasDecl>(member)) {
+        // Resolve this nested type to this type alias.
+        pa = new PotentialArchetype(this, alias);
+        
+        if (alias->hasUnderlyingType()) {
+          auto type = alias->getUnderlyingType();
+          if (type->hasArchetype()) {
+            auto repr = alias->getUnderlyingTypeLoc().getTypeRepr();
+            
+            if (auto identRepr = dyn_cast<IdentTypeRepr>(repr)) {
+              PotentialArchetype *existingPA = this;
+              for (auto component : identRepr->getComponentRange()) {
+                existingPA = existingPA->getNestedType(component->getIdentifier(),
+                                                       builder);
+                if (!existingPA)
+                  break;
+              }
+              if (!existingPA)
+                continue;
```

Comment:
It currently sets pa->ArchetypeOrConcreteType to the concrete type, which appears to work for anywhere this alias appears in a same type requirement. Is there something else I'm missing?